### PR TITLE
[DPE-4876][DPE-4943][DPE-4944] Add/Remove shards + fix `error` in secret-changed hook

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Issue
1. Charm does not support adding/remove shares (even though the dependent library does)
2. Shards briefly have their `secret-changed` hook go into `error` state on start up. This is due to the fact that we try to start `pbm` but on shards they cannot start without a config-server

## Solution
1. Make necessary changes in `src` files so that lib code for adding/removing shards works
2. Do not start pbm on shards that have not been added to a config server 

## Tests
1. Add integration tests for adding/removing shards (copied from Charmed MongoDB VM)
2. Remove `ignore_errors` in integration tests

### CODE REVIEW

File: tests/integration/sharding_tests/test_sharding.py [hunk 1]
```diff
@@ -68,5 +73,55 @@ async def test_build_and_deploy(ops_test: OpsTest) -> None:
...
+        ops_test, SHARD_TWO_APP_NAME, status=SHARD_NEEDS_CONFIG_SERVER_STATUS, timeout=300
+    )
+    await wait_for_mongodb_units_blocked(
+        ops_test, SHARD_THREE_APP_NAME, status=SHARD_NEEDS_CONFIG_SERVER_STATUS, timeout=300
+    )
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_cluster_active(ops_test: OpsTest) -> None:
```
Comment (MiaAltieri) [inline tests/integration/sharding_tests/test_sharding.py:100]: tests copied from VM charm

File: tests/integration/helpers.py [hunk 2]
```diff
@@ -519,3 +569,86 @@ def audit_log_line_sanity_check(entry) -> bool:
             logger.error("Field '%s' not found in audit log entry \"%s\"", field, entry)
             return False
     return True
+
+
+async def get_unit_hostname(ops_test: OpsTest, unit_id: int, app: str) -> str:
```
Comment (MiaAltieri) [inline tests/integration/helpers.py]: copied from VM charm

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -48,6 +51,46 @@
 logger = logging.getLogger(__name__)
 
 
+class Status:
```
Comment (MiaAltieri) [inline tests/integration/helpers.py]: copied from VM charm

File: tests/integration/sharding_tests/helpers.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+from typing import List, Optional
+
+
+def get_cluster_shards(mongos_client) -> set:
```
Comment (MiaAltieri) [inline tests/integration/sharding_tests/helpers.py:7]: copied from VM charm

File: src/charm.py [hunk 5]
```diff
@@ -430,9 +444,43 @@ def drained(self) -> bool:
...
+        for unit in self.peers_units:
+            if primary_ip == self.unit_host(unit):
+                return unit.name
+
+        return None
+
     # END: properties
 
     # BEGIN: generic helper methods
+    def remote_mongos_config(self, hosts) -> MongosConfiguration:
```
Comment (MiaAltieri) [inline src/charm.py:471]: copied from VM charm

**Review by Gu1nness — DISMISSED** [review]

File: src/charm.py [hunk 6]
```diff
@@ -344,12 +352,18 @@ def is_role(self, role_name: str) -> bool:
         return self.role == role_name
 
     def has_config_server(self) -> bool:
-        """TODO: Implement this function as part of sharding."""
-        return False
+        """Returns True if we have a config-server."""
+        return self.get_config_server_name() is not None
 
     def get_config_server_name(self) -> None:
```
Comment (Mehdi-Bendriss) [inline src/charm.py]: nit: 
```suggestion
    def get_config_server_name(self) -> str | None:
```

File: src/charm.py [hunk 7]
```diff
@@ -430,9 +444,43 @@ def drained(self) -> bool:
 
         return self.unit_peer_data.get("drained", False)
 
+    @property
+    def primary(self) -> str:
```
Comment (Mehdi-Bendriss) [inline src/charm.py]: nit: 
```suggestion
    def primary(self) -> str | None:
```

File: tests/integration/sharding_tests/helpers.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+def get_databases_for_shard(mongos_client, shard_name) -> Optional[List[str]]:
+    """Returns the databases hosted on the given shard."""
+    config_db = mongos_client["config"]
+    if "databases" not in config_db.list_collection_names():
+        return None
+
+    databases_collection = config_db["databases"]
+
+    if databases_collection is None:
```
Comment (Mehdi-Bendriss) [inline tests/integration/sharding_tests/helpers.py:36]: can it be `None` here? assuming `"databases"` I think we should perhaps:
```suggestion
    if not databases_collection:

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks Mia! I have a small question

**Review by Gu1nness — DISMISSED** [review]

**Review by Mehdi-Bendriss — DISMISSED** [review]

**Review by Mehdi-Bendriss — APPROVED** [review]

**Review by Gu1nness — APPROVED** [review]

**Pull request merged**